src/Chapter3.hs

```diff
@@ -480,6 +516,50 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Castle
+  = NoCastle
+  | Castle String
+  deriving (Show)
+
+data Wall
+  = NoWall
+  | Wall
+  deriving (Show)
+
+data AdvancedBuilding
+  = Church
+  | Library
+  deriving (Show)
+
+data House = House
+  { people :: Int }
+  deriving (Show)
+
+mkHouse :: Int -> House
+mkHouse x
+  | 1 <= x && x <= 4 = House { people = x }
+  | otherwise = House { people = (max 1 (min 4 x)) }
+
+data City = City
+  { cityCastle :: Castle
+  , cityWalls :: Wall
```

Comment:
This is the correct data type, that can represent what we have in requirements 👍🏼 
However, this also allows having some states that should not ideally exist: like having walls without any castle at all. With this representation, this is possible. However, we actually can restrict this on type-level by providing more representative types 🙂 This is actually the greatest power of Haskell and type system 😍 

This pattern is called "Make illegal states unrepresentable", and you can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable

One way to avoid it is to change the `Castle` data type:

Let's look at the task:

> Wall, but only if the city has a castle
> -- Optional castle with a name (as 'String')

That means that walls are very connected to the castle, and they could not exist separately. So we have three options for the castle:

1. Castle with a name
2. Castle with the name + walls
3. No castle

So you could add another constructor to the `Castle` type:

```haskell
data Castle 
    = NoCastle
    | Castle String
    | WallCastle String
```